R/decostand.R

```diff
@@ -269,3 +311,323 @@
         x[abs(x) < sqrt(.Machine$double.eps)] <- 0
     x
 }
+
+
+OptSpace <- function(x, ROPT=3, NITER=5, TOL=1e-5, verbose=FALSE)
+{
+
+  ## Preprocessing : x     : partially revealed matrix
+  if (is.data.frame(x)) {
+    x <- as.matrix(x)
+  }
+  if (!is.matrix(x)){
+    stop("* OptSpace : an input x should be a matrix")
+  }
+  if (any(is.infinite(x))){
+    stop("* OptSpace : no infinite value in x is allowed")
+  }
+  if (!any(is.na(x))){
+    stop("* OptSpace : there is no unobserved values as NA")
+  }
+  idxna <- (is.na(x))
+  M_E <- array(0, c(nrow(x), ncol(x)))
+  M_E[!idxna] <- x[!idxna]
+  
+  ## Preprocessing : size information
+  n <- nrow(x)
+  m <- ncol(x)
+  
+  ## Preprocessing : other sparse-related concepts
+  nnZ.E <- sum(!idxna)
+  E <- array(0, c(nrow(x), ncol(x))); E[!idxna] <- 1
+  eps <- nnZ.E/sqrt(m*n)
+  
+  ## Preprocessing : ROPT  : implied rank
+  if (is.na(ROPT)){
```

Comment:
I don't like this design; Setting an argument to be `NA` implies that its value is *missing*. Better conventions are `NULL`. However, in this case, I think we should define `ropt = TRUE` as the argument and default value. Then in the function we would do

```r
if (ropt) {
  rpot <- 3 # set your default
}
```

and then we can document `ropt` to take a logical `TRUE/FALSE` or a positive integer for the rank.